sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamMatchRel.java

```diff
@@ -0,0 +1,338 @@
+      // operators here so the sorted result will be preserved
+      // for the next match transform.
+      // In most of the runners (if not all) this should be true.
+      PCollection<KV<Row, Iterable<Row>>> orderedUpstream =
+          groupedUpstream.apply(ParDo.of(new SortPerKey(upstreamSchema, orderKeyList)));
+
+      // apply the pattern match in each partition
+      ArrayList<CEPPattern> cepPattern =
+          CEPUtil.getCEPPatternFromPattern(upstreamSchema, pattern, patternDefs);
+      String regexPattern = CEPUtil.getRegexFromPattern(pattern);
+      PCollection<KV<Row, Iterable<Row>>> matchedUpstream =
+          orderedUpstream.apply(ParDo.of(new MatchPattern(cepPattern, regexPattern)));
+
+      // apply the ParDo for the measures clause
+      // for now, output all rows of each pattern matched (for testing purpose)
+      // TODO: add ONE ROW PER MATCH and MEASURES implementation.
+      PCollection<Row> outStream =
+          matchedUpstream.apply(ParDo.of(new Measure())).setRowSchema(upstreamSchema);
+
+      return outStream;
+    }
+
+    private static class Measure extends DoFn<KV<Row, Iterable<Row>>, Row> {
+
+      @ProcessElement
+      public void processElement(@Element KV<Row, Iterable<Row>> keyRows, OutputReceiver<Row> out) {
+        for (Row i : keyRows.getValue()) {
+          out.output(i);
+        }
+      }
+    }
+
+    // TODO: support both ALL ROWS PER MATCH and ONE ROW PER MATCH.
+    // support only one row per match for now.
+    private static class MatchPattern extends DoFn<KV<Row, Iterable<Row>>, KV<Row, Iterable<Row>>> {
+
+      private final ArrayList<CEPPattern> pattern;
+      private final String regexPattern;
+
+      MatchPattern(ArrayList<CEPPattern> pattern, String regexPattern) {
+        this.pattern = pattern;
+        this.regexPattern = regexPattern;
+      }
+
+      @ProcessElement
+      public void processElement(
+          @Element KV<Row, Iterable<Row>> keyRows, OutputReceiver<KV<Row, Iterable<Row>>> out) {
+        ArrayList<Row> rows = new ArrayList<>();
+        StringBuilder patternString = new StringBuilder();
+        for (Row i : keyRows.getValue()) {
+          rows.add(i);
+          // check pattern of row i
+          String patternOfRow = " "; // a row with no matched pattern is marked by a space
+          for (int j = 0; j < pattern.size(); ++j) {
+            CEPPattern tryPattern = pattern.get(j);
+            if (tryPattern.evalRow(i)) {
+              patternOfRow = tryPattern.getPatternVar();
+            }
+          }
+          patternString.append(patternOfRow);
+        }
+
+        Pattern p = Pattern.compile(regexPattern);
+        Matcher m = p.matcher(patternString.toString());
+        // if the pattern is (A B+ C),
+        // it should return a List three rows matching A B C respectively
+        if (m.matches()) {
+          out.output(KV.of(keyRows.getKey(), rows.subList(m.start(), m.end())));
+        }
```

Comment:
I realized this should be a while loop. And for regex implementation, the default after match strategy is "skip past last row". I will change it to a while loop in my next commit.